### PR TITLE
candidate beam search : Add candidate-based response generation functionality to the generate function

### PR DESCRIPTION
**Pull request by ckdrjs96**
# What does this PR do?

This PR enhances the generate functionality by pre-defining desired responses as candidates. The GPT model will generate responses exclusively from within this candidate set. This is particularly useful for generating verified responses, feature generation, or when dealing with a large number of classifications (hundreds to thousands).

It pre-encodes the desired candidates and modifies the beam search function to exclude token IDs not present in the candidates. The scores for excluded tokens are set to -inf, ensuring that only the desired candidates are generated.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@gante

### CODE REVIEW

**Comment by gante** [comment]
Hi @ckdrjs96 👋 Thank you for opening the PR!

`transformers` is a large library, so we must be conservative when adding new features to ensure we can maintain them. To be considered for addition, a feature must be one of the following:
1. It is a very small feature, with negligible code footprint and maintenance requirements;
2. It is a feature that is widely requested by the community, ideally linking requests for the feature;
3. The feature substantially improves a certain aspect of the library and/or model usage, and there is data to back the claim.

Unfortunately, I can't fit this feature in any of the above buckets at the moment. Let me know if the situation changes, or if you have data to support your proposed feature, based on the criteria above 🤗

**Comment by ckdrjs96** [comment]
Hello @gante .

I think Candidate beam search is powerful when used as a feature generator.

When using GPT for zero-shot or with a slight finetuning to produce features, it is often difficult to obtain the desired consistent form of response.

Therefore, by utilizing this, you can obtain consistent forms of responses without additional training or with slight finetuning.

1. Often, it is difficult to obtain the desired response without fine-tuning when receiving a prompt.
    1. For instance, if I request a response of 'yes or no' from GPT, it might answer yes, true, no, false, and so on. However, using candidate beam search can only provide answers of yes or no.
    2. It is useful when connecting specific words to the desired space, such as with Hard verbalizer.
        1. In the case of sentiment analysis, I want to receive a response like "Ihappy," but GPT can provide very similar synonyms such as cheerful, delighted, in a good mood, glad, content, and so on.
```python
keyword_list = ['happy','sad', 'angry','love' ...]
candidates_words_ids = tokenizer(keyword_list, add_special_tokens=False,).input_ids

outputs = model.generate(
    **input_ids,
    candidate_words_ids=candidates_words_ids,
    num_beams=3,
    num_return_sequences=3,
    remove_invalid_values=True,
    max_length=512,
    eos_token_id = tokenizer.eos_token_id,
    return_dict_in_generate=True,
    output_scores=True
)
```
2. It is known that for classification, it is recommended to create embeddings with GPT and utilize softmax. However, in cases of too many classification problems, after tuning with a small dataset (about 1 to 5 labels per label), you can use candidate beam search to obtain the desired response. (The existing generation often answers with labels that have not been learned.)
    - For example, when creating approximately 1000 topic classifiers with various depths, you can store those topics in JSON and load them for use in the desired format.
    - `keyword_list = ['sports>football',''sports>baseball','movie>action', 'movie>comedy', ...] `
3. Building a trustworthy chatbot
    1. GPT often hallucinates. However, companies want safe responses.
    2. After implementing reviewed response lists in candidates, you can always receive verified responses when generating.

**Comment by gante** [comment]
@ckdrjs96 isn't it similar to libraries like https://github.com/outlines-dev/outlines ?

**Comment by ckdrjs96** [comment]
yes. It looks like the library includes all the features I need. Thanks for letting me know.

**Comment by gante** [comment]
@ckdrjs96 perfect 💛 We prefer to support external libraries, as opposed to adding similar features in `transformers` -- community power 💪
